### PR TITLE
autograd: fix PyObject* refcount leak in setup_context path

### LINKED ISSUE #188034: [BUG] Reference leak in autograd Function.apply — setup_context return value never released
## Bug Description

In `torch/csrc/autograd/python_function.cpp`, when calling a custom autograd Function's `setup_context`, the return value from `PyObject_CallObject` is stored in a raw `PyObject*` (via `auto` deduction) and never `Py_DECREF`'d on the success path. This leaks one reference per forward pass.

## Location

`torch/csrc/autograd/python_function.cpp:1610-1614`

## Code

```cpp
THPObjectPtr setup_context_fn(
    PyObject_GetAttrString(cls, "setup_context"));
auto result =
    PyObject_CallObject(setup_context_fn, ctx_input_output_tuple);
if (!result) {
    return nullptr;
}
// 'result' is PyObject* (new reference) — never Py_DECREF'd!
// Falls through to the next branch
```

`auto` deduces to `PyObject*`, not `THPObjectPtr`. On the success path (result is non-NULL), the code falls through without releasing the reference. Since `setup_context` typically returns `None`, this leaks the `None` singleton's refcount by +1 per call (masked on 3.12+ by immortalization, but still a real leak on earlier versions and a contract violation).

## Impact

Every `autograd.Function.apply()` call with a `setup_context` override leaks one reference. For training loops calling custom autograd functions millions of times, this causes unbounded memory growth on Python < 3.12. On 3.12+ it's masked by immortal objects but remains a C API contract violation.

## Suggested Fix

Change `auto` to `THPObjectPtr`:
```cpp
THPObjectPtr result(
    PyObject_CallObject(setup_context_fn, ctx_input_output_tuple));
if (!result) {
    return nullptr;
}
```

## Environment

- PyTorch version: current main
- Affected: any custom `torch.autograd.Function` with `setup_context` override

cc @ezyang @gchanan @kadeng @msaroufim @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

### PR DESCRIPTION
**Pull request by aperson30**
## Summary

- `python_function.cpp`: In `_call_function_forward`, the return value of `PyObject_CallObject` for the `setup_context` call was stored as `auto result` (raw `PyObject*`), so `Py_DECREF` was never called on the success path, leaking one reference per invocation.
- Changed declaration to `THPObjectPtr result(...)` for RAII cleanup, consistent with every other `PyObject_CallObject` call in the same file.
- Added a regression test that returns a sentinel object from `setup_context` and asserts `sys.getrefcount` is unchanged after repeated forward+backward passes.

Fixes #188034

## Test plan

- [ ] `python test/test_autograd.py TestAutograd.test_custom_function_setup_context_no_refcount_leak`
- [ ] Existing `test_custom_function_setup_context_*` tests pass

### CODE REVIEW

**Comment by aperson30** [comment]
@alanhdu @soulitzer All CI tests are green. Could someone please take a look and merge when ready? Happy to address any feedback.
